### PR TITLE
Validate request line length before checking the version

### PR DESCRIPTION
**Pull request by campersau**
- [X] You've read the [Contributor Guide](https://github.com/dotnet/aspnetcore/blob/main/CONTRIBUTING.md) and [Code of Conduct](https://github.com/dotnet/aspnetcore/blob/main/CODE-OF-CONDUCT.md).
- [X] You've included unit or integration tests for your change, where applicable.
- [x] You've included inline docs for your change, where applicable.
- [X] There's an open issue for the PR that you are making. If you'd like to propose a new feature or change, please open an issue to discuss the change or find an existing issue.

**PR Title**
Reject request line earlier when it is not the expected length before checking for a valid version.

**PR Description**
Moves the check if the request line has the expected length before the version parsing so that it will result in an 400 (bad request) instead of a 505 status code.

By doing that the checks in `GetKnownVersionAndConfirmCR` could be removed because they are now redundant (but are still used in some tests):

https://github.com/dotnet/aspnetcore/blob/c8d548c20120ec711706fdd78d6e0e015792df54/src/Servers/Kestrel/Core/src/Internal/Infrastructure/HttpUtilities.cs#L325-L349

Addresses #33769

### CODE REVIEW

**Comment by Tratcher** [comment]
Some of the new or modified tests are failing:
https://dev.azure.com/dnceng/public/_build/results?buildId=1204486&view=ms.vss-test-web.build-test-results-tab&runId=36047970&resultId=106127&paneView=debug

File: src/Servers/Kestrel/Core/src/Internal/Infrastructure/HttpUtilities.cs [hunk 1]
```diff
@@ -320,29 +324,20 @@ public static bool GetKnownVersion(this ReadOnlySpan<byte> span, out HttpVersion
...
         /// To optimize performance the HTTP/1.1 will be checked first.
         /// </remarks>
-        /// <returns><c>true</c> if the input matches a known string, <c>false</c> otherwise.</returns>
+        /// <returns>the HTTP version if the input matches a known string, <c>Unknown</c> otherwise.</returns>
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        internal static HttpVersion GetKnownVersionAndConfirmCR(this ReadOnlySpan<byte> location)
+        internal static HttpVersion GetKnownVersion(this ReadOnlySpan<byte> location)
         {
-            if (location.Length < sizeof(ulong))
+            Debug.Assert(location.Length > sizeof(ulong) && location[sizeof(ulong)] == (byte)'\r');
```
Comment (halter73) [inline src/Servers/Kestrel/Core/src/Internal/Infrastructure/HttpUtilities.cs:331]: This seems scary to me. Can we rename it to `GetKnownVersionUnchecked()` or something similar?

**Review by halter73 — APPROVED** [review]

File: src/Servers/Kestrel/Core/src/Internal/Infrastructure/HttpUtilities.cs (hunk 1 above)
Comment (campersau) [reply src/Servers/Kestrel/Core/src/Internal/Infrastructure/HttpUtilities.cs:331]: I have renamed the mehtod to `GetKnownVersionUnchecked`.

**Comment by Tratcher** [comment]
Looks ready to merge?

**Comment by campersau** [comment]
Yes, from my side it is. I changed the `GetKnownVersion` to use `TryReadUInt64LittleEndian` instead so it is similar to `GetKnownHttpScheme` and only does one thing.

**Comment by halter73** [comment]
Thanks! @campersau

**Pull request merged**